### ISSUE TITLE
Director does not acquire GIL before executing Python code

### ISSUE THREAD

**Issue by peastman** [swig/swig#2670]
I use SWIG to wrap a C++ library in Python.  I'm trying to use a director class to allow a method of a C++ class to be implemented in Python.  Whenever the C++ code tries to call a method written in Python, it segfaults.  I finally figured out this is because the director does not acquire the GIL before trying to run the Python code.

1. Is this a bug?  It seems like one to me.  The director should ensure it has the GIL before trying to enter the Python interpreter.
2. Is there a workaround?
3. If this is not considered a bug, could the documentation be updated to discuss it?  The current documentation on directors in Python doesn't say anything about threading issues.

**Comment by ojwb** [comment]
Are you running SWIG with `-threads`?

**Comment by peastman** [comment]
No.  But the C++ library might call the director from any thread, not necessarily one created by Python.

**Comment by peastman** [comment]
Is there a workaround?  For example, is there a way I can make SWIG insert a call to `PyGILState_Ensure()` before entering the Python method?

**Comment by ojwb** [comment]
> is there a way I can make SWIG insert a call to `PyGILState_Ensure()` before entering the Python method?

Um, that's what `-threads` does...

**Comment by peastman** [comment]
I believe the effect of `-threads` is to cause it to release the GIL when calling from Python into C++.  That's different from what I'm talking about.

**Comment by ojwb** [comment]
It should do both.

Perhaps you should actually try it rather than disbelieving advice from someone who has contributed to its implementation?

**Comment by peastman** [comment]
I don't want it to release the GIL when calling into C++.  That isn't what I'm asking how to do.  Is there a way to make it acquire the GIL when a C++ thread calls into Python *without* making every call from Python into C++ release it?

**Comment by ojwb** [comment]
The manual suggests `-threads` plus `%nothreadallow;` should do that, though I've never tried it myself.
